Matchinfohmp

gvolpe79 opened the pull request.

eflatlan committed:
Initial commit of calibration folder

Compiles on CC7, datatypes etc to be checked.
Not up to date with workflow

eflatlan committed:
Changed CcdbObjectInfo names, added get-methods

eflatlan committed:
Added hmpid-dcs-data-workflow.cxx

eflatlan committed:
HMPIDDataProcessorSpec.h:: much change needed 

DataProcessorSpec getHMPIDDCSDataProcessorSpec()
     // NB! probably change Lifetime::Sporadic

eflatlan committed:
changed methods for first/last timestamps 

evaluate first and last timestamps for RefIndex and ChargeCut in finalize<DP-identifier> methods

eflatlan committed:
changed methods for first/last timestamps 

evaluate first and last timestamps for RefIndex and ChargeCut in finalize<DP-identifier> methods

eflatlan committed:
Add files via upload

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSTime.cxx

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
added struct mTimeArNmean with values of Timestamps

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
added getRefIndexObj, getChargeCutObj

	const HMPIDThreshVars& getChargeCutObj() const { return mChargeCut; }
		//HMPIDThreshVars mChargeThresh; in private

const HMPIDRefIndexVars& getRefIndexObj() const { return mRefIndex; }
		//HMPIDRefIndexVars mRefIndex; in private

eflatlan committed:
sendRefIndexOutput sendChargeThresOutput

added Qs

eflatlan committed:
Qs about structs 

added Qs

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
HMPIDDataProcessorSpec: header file w/o methods

header file w/o methods

eflatlan committed:
HMPIDDataProcessorSpec.cxx: implemented methods from h_

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
added #pragma link C++ class TF1[] + ;

eflatlan committed:
HMPIDDataProcessorSpec.h / cxx fault

HMPIDDataProcessorSpec.h must be linked in "hmpid-dcs-data-workflow"


Must fix : 

Detectors/HMPID/calibration/testWorkflow/hmpid-dcs-data-workflow.cxx:40: undefined reference to `o2::framework::getHMPIDDCSDataProcessorSpec()'

eflatlan committed:
Update HMPIDDataProcessorSpec.h

eflatlan committed:
Added HMPIDDCSDataProcessorSpec, compiles

Strings of aliases still need to be changed,

eflatlan committed:
fixed links

eflatlan committed:
fixed var and function names

eflatlan committed:
Add files via upload

eflatlan committed:
Add files via upload

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.h

eflatlan committed:
Delete HMPIDDCSProcessorSpec.cxx

eflatlan committed:
Delete HMPIDDataProcessorSpec.cxx

eflatlan committed:
Delete HMPIDCalibrationLinkDef.h~

eflatlan committed:
Delete HMPIDDCSTime.h~

eflatlan committed:
Delete HMPIDDataFormats.h

eflatlan committed:
Delete hmpid-dcs-data-workflow.cxx~

eflatlan committed:
Delete hmpid-dcs-sim-workflow.cxx~

eflatlan committed:
Delete _HMPIDDataProcessorSpec.h

eflatlan committed:
Delete _HMPIDDataProcessorSpec.cxx

eflatlan committed:
Delete _HMPIDDCSDataProcessorSpec.h

eflatlan committed:
Delete HMPIDDataProcessorSpec.h_

eflatlan committed:
Delete HMPIDDataProcessorSpec.h

eflatlan committed:
Delete HMPIDDataProcessorSpec.cxx

eflatlan committed:
Delete HMPIDDCSDataProcessorSpec.h

eflatlan committed:
removed unused files

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
String aliases for HMPID

eflatlan committed:
add hardcoded alias-strings 

is using a for loop ok, or should use something like: 
https://github.com/AliceO2Group/AliceO2/blob/f76eb23c3fa8d81c5bfea9071d047ef548059fd4/Detectors/ITSMFT/MFT/condition/testWorkflow/MFTDCSDataProcessorSpec.h#L92-L115

eflatlan committed:
moved alias-strings to header-file

eflatlan committed:
moved alias-strings to header-file

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.h

eflatlan committed:
updated datatype checks

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Add files via upload

eflatlan committed:
Add files via upload

eflatlan committed:
Create makeHMPIDCCDBEntryForDCS.C

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.h

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
Add files via upload

eflatlan committed:
Add files via upload

eflatlan committed:
Delete makeHMPIDCCDBEntryForDCS.C

eflatlan committed:
Update HMPIDCalibrationLinkDef.h

eflatlan committed:
Update CMakeLists.txt

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Create HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
changed subStringToInt function

eflatlan committed:
Add files via upload

eflatlan committed:
Update makeHMPIDCCDBEntryForDCS.C

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
added checks for ranges of ch, sec, rad numbers

eflatlan committed:
remove unnecessary vars

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.h

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Update hmpid-dcs-sim-workflow.cxx

eflatlan committed:
Update hmpid-dcs-sim-workflow.cxx

eflatlan committed:
Rename checkHMPIDAliasStrings.C to checkHMPIDAliasStrings.txt

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.cxx

eflatlan committed:
Update HMPIDCalibrationLinkDef.h

eflatlan committed:
Update hmpid-dcs-sim-workflow.cxx

eflatlan committed:
Update hmpid-dcs-sim-workflow.cxx

eflatlan committed:
Add files via upload

eflatlan committed:
add_subdirectory(calibration) in /HMPID

eflatlan committed:
add if(BUILD_TESTING)  

To add makeCCDBHMPIDEntryforDCS:
if(BUILD_TESTING) 
  add_subdirectory(calibration/macros)
 endif()

eflatlan committed:
upload strings from DP-hints

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
Changed to vector<vector <TF1*>>

Changed CCDB objects to  vector<vector <TF1*>> from arrays of TF1-vectors

command o2-hmpid-dcs-workflow does not run

eflatlan committed:
changed to smart-pointers

eflatlan committed:
Update HMPIDDCSDataProcessorSpec.h

eflatlan committed:
Update HMPIDDCSProcessor.h

eflatlan committed:
lowercase procTrans and defaultEMean

eflatlan committed:
lowercase procTrans and defaultEMEan

eflatlan committed:
change formatting, change dpcom-vector names

eflatlan committed:
Update HMPIDDCSProcessor.cxx

eflatlan committed:
format to advised indent

eflatlan committed:
format to guidelines

eflatlan committed:
format to guidelines

eflatlan committed:
format to guidelines

eflatlan committed:
format to guidelines

eflatlan committed:
Update HMPIDCalibrationLinkDef.h

eflatlan committed:
format to guidelines

eflatlan committed:
format to guidelines

Change functions and variables to start with lowercase

change indent to 2 spaces

eflatlan committed:
reformat to guidelines

Correct name-convention for functions and vars, 
correct indentations

gvolpe79 committed:
Merge branch 'AliceO2Group:dev' into dev

eflatlan committed:
Add files via upload

Commit:
Fixing commits

gvolpe79 committed:
Merge branch 'AliceO2Group:dev' into dev

Commit:
Fixing commts 2

Commit:
Matching info for the HMPID implemented and added to the RecoContainer

Commit:
Fixing previous commit

sawenzel commented:
@gvolpe79 : Would you be able to provide a reasonable description for this PR? We should also squash the ~130 commits most likely.

gvolpe79 commented:
Hi Sandro, 
I created the matching info for the HMPID, and modified the RecoContainer adding the HMPID matching info and the clusters. 
There was anyway a problem! The student we have here at cern, that is doing some development in parallel for the HMPID calibration, committed directly on the dev of my fork end not in a separate branch as I requested. Once I realized this, I updated the dev removing all his commits and started from there for the matchinfohmp branch. I think it should be ok.

gvolpe79 committed:
Added trigger information

shahor02 commented on DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h:
```diff
@@ -213,6 +220,7 @@ struct DataRequest {
   void requestTRDTracklets(bool mc);
   void requestMCHClusters(bool mc);
   void requestMIDClusters(bool mc);
+  void requestHMPClusters(bool mc);
```
You should also add `requestHMPMatches(bool mc)`

shahor02 commented on DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h:
```diff
@@ -580,6 +595,11 @@ struct RecoContainer {
 
+  // HMPID clusters
+  auto getHMPClusters() const { return getSpan<o2::hmpid::Cluster>(GTrackID::HMP, CLUSTERS); }
+  auto getHMPTriggers() const { return getSpan<o2::hmpid::Trigger>(GTrackID::HMP, CLUSREFS); }
```
Your matches are also grouped per trigger, right? Then there should be separate 
```
auto getHMPClusterTriggers() const { return getSpan<o2::hmpid::Trigger>(GTrackID::HMP, CLUSREFS); }
auto getHMPClusters() const { return getSpan<o2::hmpid::Cluster>(GTrackID::HMP, CLUSTERS); }
```
and e.g.
```
auto getHMPMatchTriggers() const { return getSpan<o2::hmpid::Trigger>(GTrackID::HMP, TRACKREFS); }
auto getHMPMatches() const { return getSpan<o2::hmpid::MatchInfoHMP>(GTrackID::HMP, MATCHES); }
```

gvolpe79 replied:
ok, and  I have to keep this then?

const o2::dataformats::MatchInfoHMP& getHMPMatch(GTrackID id) const { return getObject<o2::dataformats::MatchInfoHMP>(id, MATCHES); }

shahor02 replied:
As you prefer, the same can be obtained as 
```
auto mtspan = recoCont.getHMPMatches();
auto& match = mtspan[id.getIndex()];
```
Note that each `getHMPMatch(GTrackID id) ` would go via new span creation. It is not very expensive but still more expensive than getting span once then iterating over its elements.

shahor02 commented on DataFormats/Reconstruction/include/ReconstructionDataFormats/MatchInfoHMP.h:
```diff
@@ -0,0 +1,69 @@
+  float mMipCluQ = 0.0;                 // MIP cluster charge
+  int mMipCluSize = 0.0;                // MIP cluster size
+  int mNPhots = 0.0;                    // number of candidate photo-electrons
+  std::vector<float>* mPhotsClusCharge; // list of cluster charge of candidate photo-electrons
```
we are trying to avoid dynamical size data formats, moreover with raw pointer, such objects are not messageable via shared memory and need require expensive root streaming. 
Do you really need a vector of charges? In such cases we usually add yet another vector which would keep the charges and the match would just keep the 1st entry index and the number of entries it has in this vector.

gvolpe79 replied:
It refers to the charge of the cherenkov photon clusters for the given track. This is a information very useful and I would like to not lose it! So adding an other vector, but it is not clear to me where?

shahor02 replied:
E.g. HMPClusterCharges via
```
auto getHMPClusterCharges() const { return getSpan<float>(GTrackID::HMP, PATTERNS); }
```
The idea is to refer to the range of entries in this vector from the HMPCluste in the same way as you refer from HMPTriggers to HMPClusters.

gvolpe79 committed:
Added triggers and photon cluster charge

shahor02 commented on DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx:
```diff
@@ -266,6 +266,17 @@ void DataRequest::requestMCHClusters(bool mc)
+void DataRequest::requestHMPClusters(bool mc)
+{
+  addInput({"hmpidcluster", "HMP", "CLUSTERS", 0, Lifetime::Timeframe});
+  addInput({"hmpidtriggers", "HMP", "CLUSREFS", 0, Lifetime::Timeframe});
```
The special vector of charges is part of the clusters info, should be added here

shahor02 replied:
discard this comment, it was wrong

shahor02 commented on DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx:
```diff
@@ -941,6 +967,15 @@ void RecoContainer::addTOFClusters(ProcessingContext& pc, bool mc)
+void RecoContainer::addHMPClusters(ProcessingContext& pc, bool mc)
+{
+  commonPool[GTrackID::HMP].registerContainer(pc.inputs().get<gsl::span<o2::hmpid::Cluster>>("hmpidcluster"), CLUSTERS);
+  commonPool[GTrackID::HMP].registerContainer(pc.inputs().get<gsl::span<o2::hmpid::Trigger>>("hmpidtriggers"), CLUSREFS);
```
here also the cluster charges vector must be added.

shahor02 replied:
discard this comment, it was wrong

shahor02 commented on DataFormats/Detectors/HMPID/include/DataFormatsHMP/Cluster.h:
```diff
@@ -120,6 +120,16 @@ class Cluster
 };
 
 } // namespace hmpid
+
```
please fix the Cluster class, it should not contain the pointer on the vector, instead the charges should be accessed from the separate vector.

shahor02 replied:
discard this comment, it was wrong

shahor02 reviewed (changes requested):
Hi @gvolpe79 

See comments below. In short: your clusters still allocate the memory via pointer to the vector, they should not (they are not messageable now, even if you declare them to be so). The variable size info should come via separate vector, the cluster will refer to the entries in this vector.
For the matches info (which should go the AOD together with matched clusters info?): it should either contain to the selected clusters vector which is also stored together with refilled charges vector or (preferable to not increase the number of vectors) refer directly on the charges vector of selected clusters.
In any case, if I understood correctly what you need, there will be 2 vector of charges: one produced together with clusters and another one only for charges of matched clusters

shahor02 reviewed (changes requested):
Hi @gvolpe79 

Please discard my previous comments, they were wrong:  I've probably misread your Cluster class, thinking that the mDigs vector you have there is also for the charges. Now I realize that you have if only at the clustering stage and then don't use it, is it so?

But there are another problems:
1) you declared/implemented `requestHMPClusters()` but you should invoke it in from the `requestClusters`, e.g. here (see the same for other detectors):
https://github.com/AliceO2Group/AliceO2/blob/183cdd3e763b2197bc63d832e25467e3b835614a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L512

2) You have declared `requestHMPMathes` but you neither implement it nor invoked it from the  requestTracks, should add somewhere here:
https://github.com/AliceO2Group/AliceO2/blob/183cdd3e763b2197bc63d832e25467e3b835614a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L475

3) You have declared/implemented `addHMPMatches` but you should also invoke it e.g. in https://github.com/AliceO2Group/AliceO2/blob/183cdd3e763b2197bc63d832e25467e3b835614a/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L701

Commit:
Updating RecoContainer.cxx

gvolpe79 commented:
Hi @shahor02, 
I just committed the requested modifications!

shahor02 committed:
Add getter for HMP matches labels

shahor02 committed:
dummy commit to retrigger CI
